src/component.jl

```diff
@@ -66,12 +66,12 @@ function Base.permute!(c::AbstractComponent, permvec)
     permute!(c.indices, permvec)
 end
 
-Base.@propagate_inbounds @inline Base.getindex(c::Component, e::Entity) = c.data[c.indices[e.id]]
-Base.@propagate_inbounds @inline Base.getindex(c::SharedComponent, e::Entity) = c.shared[c.data[c.indices[e.id]]]
-Base.@propagate_inbounds @inline Base.getindex(c::Component, i::Integer) = c.data[i]
+Base.@propagate_inbounds @inline Base.getindex(c::Component,       e::AbstractEntity)  = c.data[c.indices[e.id]]
+Base.@propagate_inbounds @inline Base.getindex(c::SharedComponent, e::AbstractEntity)  = c.shared[c.data[c.indices[e.id]]]
```

Comment:
Yes the getproperty overloading made it easy to get working with previous functions, just changing ::Entity to ::AbstractEntity. What a nice feature that is

Reply:
Yeah it's a fantastic feature. I think it's fine to rely on here as long as you don't expect users to make their own `AbstractEntity` types (which seems unlikely?)

Reply:
Well, or the interface requires it to have `.id`.
Actually, this is something I've wondered a bit about, there's no way of explicitly requiring certain things from types that subtype a give abstract type is there? Basically the requirements need to be documented and people wanting to use the functionality need to implement it. 

There used to be this way of doing things with
```julia
Entity(e::Entity) = e
id(e::AbstractEntity) = Entity(e).id
...
```
so that anyone wanting to use all the code with `AbstractEntity` only had to implement the `Entity(e::NewEntity)` function and everything worked. Is there a "best practices"  way related to overloading `getproperty`?

Reply:
> Basically the requirements need to be documented and people wanting to use the functionality need to implement it.

Yes that's exactly right.

> Is there a "best practices" way related to overloading getproperty?

I'd say that `getproperty` is unusual in that
* It rarely makes sense to provide an implementation for abstract types.
* It's fiddly to implement correctly for concrete types, and also have convenient access to the internal fields

So overall, I feel that `getproperty` is not that good for building generic interfaces, but it's great for providing user-visible *convenience interfaces* for concrete types.

```julia
id(e::AbstractEntity) = Entity(e).id
```

I'd just turn this around, and require that people implement `id` as part of defining their custom entity type.

```julia
id(e::Entity) = e.id
id(e::EntitytState) = id(e.e)
```

Reply:
Yes, this is also possible in this case, since it's a small concrete type anyway. In general the `Entity(e).id` thing is nice for larger types so that if there are many fields used by the package code, users don't need to implement accessors to each of those, just the `Entity` method.